pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -34,87 +34,127 @@ import (
+	gc        *GarbageCollector
 }
 
-func (p *Propagator) addToGraphLookup(n *node) {
+func (p *Propagator) insertNode(n *node) {
 	fmt.Println("CHAO: add to graph:", n.identity)
-	p.graphLookup[n.identity.UID] = n
+	p.uidToNode[n.identity.UID] = n
 }
 
-func (p *Propagator) removeFromGraphLookup(n *node) {
+func (p *Propagator) removeNode(n *node) {
 	fmt.Println("CHAO: remove from graph:", n.identity)
-	delete(p.graphLookup, n.identity.UID)
+	delete(p.uidToNode, n.identity.UID)
 }
 
-func (p *Propagator) updateOwners(n *node) (ownerExists bool) {
-	fmt.Println("CHAO: in updateOwners, graphLookUp:", p.graphLookup)
+// addToOwners finds all owners as listed in n.owners, and marks n as one of
+// their dependents. ownerExists indicates if any owner listed in n.owners exist
+// in p.uidToNode.
+func (p *Propagator) addToOwners(n *node) (ownerExists bool) {
+	fmt.Println("CHAO: in addToOwners, uidToNode:", p.uidToNode)
 	for _, owner := range n.owners {
-		fmt.Println("CHAO: in updateOwners, lookup for:", owner.UID)
-		ownerNode, ok := p.graphLookup[owner.UID]
+		fmt.Println("CHAO: in addToOwners, lookup for:", owner.UID)
+		ownerNode, ok := p.uidToNode[owner.UID]
 		if !ok {
 			fmt.Println("CHAO: lookup not found")
 			continue
 		}
 		fmt.Println("CHAO: lookup found")
 		ownerExists = true
-		ownerNode.dependents = append(ownerNode.dependents, n)
+		ownerNode.dependents[n] = struct{}{}
 	}
 	return ownerExists
 }
 
-// TODO: finish this function.
-func equalReferences(a []metatypes.OwnerReference, b []metatypes.OwnerReference) bool {
-	if len(a) != len(b) {
-		return false
+// removeFromOwners finds all owners as listed in n.owners, and removes n from
+// their dependents list.
+func (p *Propagator) removeFromOwners(n *node) {
+	fmt.Println("CHAO: in removeFromOwners, uidToNode:", p.uidToNode)
+	for _, owner := range n.owners {
+		fmt.Println("CHAO: in removeFromOwners, lookup for:", owner.UID)
+		ownerNode, ok := p.uidToNode[owner.UID]
+		if !ok {
+			fmt.Println("CHAO: lookup not found")
+			continue
+		}
+		fmt.Println("CHAO: lookup found")
+		delete(ownerNode.dependents, n)
 	}
-	// sort and compare
-	return true
+}
+
+func ReferencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerReference) (added []metatypes.OwnerReference, removed []metatypes.OwnerReference) {
+	oldUIDToRef := make(map[string]metatypes.OwnerReference)
+	for i := 0; i < len(old); i++ {
+		oldUIDToRef[string(old[i].UID)] = old[i]
+	}
+	oldUIDSet := sets.StringKeySet(oldUIDToRef)
+	newUIDToRef := make(map[string]metatypes.OwnerReference)
+	for i := 0; i < len(new); i++ {
+		newUIDToRef[string(new[i].UID)] = new[i]
+	}
+	newUIDSet := sets.StringKeySet(newUIDToRef)
+
+	addedUID := newUIDSet.Difference(oldUIDSet)
+	removedUID := oldUIDSet.Difference(newUIDSet)
+
+	for uid, _ := range addedUID {
+		added = append(added, newUIDToRef[uid])
+	}
+	for uid, _ := range removedUID {
+		removed = append(removed, oldUIDToRef[uid])
+	}
+	return added, removed
```

Comment:
We may want to profile this function once everything else is working-- the naive N^2 algorithm is probably better when N is 1 or 2, since it doesn't need to allocate the maps.